src/arkode/arkode_arkstep_io.c

```diff
@@ -677,12 +679,66 @@ int arkStep_SetDefaults(ARKodeMem ark_mem)
   step_mem->msbp     = MSBP;  /* max steps between updates to J or P */
   step_mem->stages   = 0;     /* no stages */
   step_mem->istage   = 0;     /* current stage */
-  step_mem->Be       = NULL;  /* no Butcher tables */
-  step_mem->Bi       = NULL;
-  step_mem->NLS      = NULL; /* no nonlinear solver object */
   step_mem->jcur     = SUNFALSE;
   step_mem->convfail = ARK_NO_FAILURES;
   step_mem->stage_predict = NULL; /* no user-supplied stage predictor */
+
```

Comment:
Is it assumed that `ReInit` is must be called after `SetDefautls`? Otherwise I think this leaves the integrator in an invalid state i.e., `NULL` tables and nonlinear solver. What should be done with the linear solver?

If `step_mem->explicit` and `step_mem->implicit` above are reset based on `fe` and `fi`, then the default tables and a nonlinear solver could be attached below (like the controller). However, this could still leave the integrator in a bad state e.g., if a fixed-point solver was attached previously and this attaches the default Newton solver it would not have a linear solver.

Maybe this should leave the current methods and solvers in place and only reset integrator parameters. In which case some values above should be retained e.g., `step_mem->explicit`, `step_mem->implicit`, `step_mem->stages`, etc.

Reply:
The modifications to this file change nothing about any assumed order of operations regarding `ReInit` and `SetDefaults` -- the previous version merely nullified the pointer *without ever checking whether memory should be freed*.  This was @Steven-Roberts's main point in issuing the bug report, and that behavior has now been fixed.

If the previous implementation (and now current implementation) assume a specific order of operations, and if you would like to revisit those assumptions, I'd prefer that this question be handled in a separate PR.

Reply:
Agreed, this is a separate problem from the memory leak (I've opened a new issue) and can be addressed in a separate PR.